pushsync/pusher.go

```diff
@@ -0,0 +1,343 @@
+	var chunks <-chan chunk.Chunk
+	var unsubscribe func()
+	var syncedAddrs []storage.Address
+	var syncedItems []*pushedItem
+
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	// register handler for pssReceiptTopic on pss pubsub
+	deregister := p.ps.Register(pssReceiptTopic, false, func(msg []byte, _ *p2p.Peer) error {
+		return p.handleReceiptMsg(msg)
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+	ctx := context.Background()
+
+	var average uint64 = 1000
+	var measurements uint64
+
+	for {
+		select {
+
+		// retry interval timer triggers starting from new
+		case <-timer.C:
+			// initially timer is set to go off as well as every time we hit the end of push index
+			// so no wait for retryInterval needed to set  items synced
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// if subscribe was running, stop it
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+
+			log.Debug("set chunk status to synced, insert to db GC index")
+			syncedTags := make(map[uint32]int)
+			// set chunk status to synced, insert to db GC index
+			if err := p.store.Set(ctx, chunk.ModeSetSync, syncedAddrs...); err != nil {
+				log.Error("error setting chunks to synced", "err", err)
+				continue
+			}
+			for i, item := range syncedItems {
+				// increment synced count for the tag if exists
+				tag := item.tag
+				if tag != nil {
+					syncedTags[tag.Uid] = syncedTags[tag.Uid] + 1
+					item.span.Finish()
+				}
+				delete(p.pushed, syncedAddrs[i].Hex())
+			}
+			// iterate over tags in this batch
+			for uid, n := range syncedTags {
+				tag, _ := p.tags.Get(uid)
+				tag.IncN(chunk.StateSynced, n)
+				if tag.Done(chunk.StateSynced) {
+					log.Info("closing root span for tag", "taguid", tag.Uid, "tagname", tag.Name)
+					tag.FinishRootSpan()
+				}
+			}
+			// reset synced list
+			syncedAddrs = nil
+			syncedItems = nil
+
+			// we don't want to record the first iteration
+			if chunksInBatch != -1 {
+				// this measurement is not a timer, but we want a histogram, so it fits the data structure
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
+				metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
+			}
+			chunksInBatch = 0
+			batchStartTime = time.Now()
+
+			// and start iterating on Push index from the beginning
+			chunks, unsubscribe = p.store.SubscribePush(ctx)
+			// reset timer to go off after retryInterval
+			timer.Reset(retryInterval)
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				timer.Reset(0)
```

Comment:
shouldn't you be calling `unsubscribe()` here too? since every time you create a new `chunks` channel, you'd probably want to unsubscribe it once you dump the `chunks` channel